src/endpoint/s3/ops/s3_post_object_select.js

```diff
@@ -0,0 +1,185 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const dbg = require('../../../util/debug_module')(__filename);
+const S3Error = require('../s3_errors').S3Error;
+const s3_utils = require('../s3_utils');
+const http_utils = require('../../../util/http_utils');
+const { Transform } = require('readable-stream');
+const nb_native = require('../../../util/nb_native');
+
+class S3SelectStream extends Transform {
+
+    constructor(opts) {
+        super(opts);
+        this.CephS3lect = new (nb_native().CephS3lect)(opts);
+    }
+
+    async _transform(chunk, encoding, cb) {
+        this.push(await this.CephS3lect.write(chunk));
+        return cb();
+    }
+
+    async _flush(cb) {
+        this.push(await this.CephS3lect.flush());
+        return cb();
+    }
+}
+
+class EchoDebugStream extends Transform {
+
+    constructor(opts) {
+        super(opts);
+        this.header = opts.header;
+    }
+
+    _transform(chunk, enc, cb) {
+        dbg.log1(this.header, chunk.toString());
+        this.push(chunk);
+        return cb();
+    }
+}
+
+var end_message = Buffer.concat([
+    Buffer.from([0,0,0,56]), //total byte length
+    Buffer.from([0,0,0,40]), //headers byte length
+    Buffer.from([0xc1,0xc6,0x84,0xd4]), //crc of above
+    Buffer.from([11]), //byte length of below header name
+    Buffer.from(':event-type'),
+    Buffer.from([7,0,3]), //type (7, string) of below header value
+    Buffer.from('End'),
+    Buffer.from([13]), //again for second header
+    Buffer.from(':message-type'),
+    Buffer.from([7,0,5]),
+    Buffer.from('event'),
+    Buffer.from([0xfe, 0x2c, 0xee, 0x99]) //crc of above
+]);
+
+/**
+ * https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html
+ */
+async function post_object_select(req, res) {
+
+    req.object_sdk.setup_abort_controller(req, res);
+    const agent_header = req.headers['user-agent'];
+    const noobaa_trigger_agent = agent_header && agent_header.includes('exec-env/NOOBAA_FUNCTION');
+    const encryption = s3_utils.parse_encryption(req);
+    let part_number;
+    // If set, part_number should be positive integer from 1 to 10000
+    if (req.query.partNumber) {
+        part_number = s3_utils.parse_part_number(req.query.partNumber, S3Error.InvalidArgument);
+    }
+    const md_conditions = http_utils.get_md_conditions(req);
+
+    const md_params = {
+        bucket: req.params.bucket,
+        key: req.params.key,
+        version_id: req.query.versionId,
+        md_conditions,
+        encryption,
+    };
+    if (req.query.get_from_cache !== undefined) {
+        md_params.get_from_cache = true;
+    }
+    if (part_number) {
+        md_params.part_number = part_number;
+    }
+    const object_md = await req.object_sdk.read_object_md(md_params);
+
+    s3_utils.set_response_object_md(res, object_md);
```

Comment:
is this correct?? adding these response headers doesn't seem related to this op.